File: packages/gbnf-experiment/data/20260909T002206Z_273cf6b9/ported_implementation/tests/unit/grammar-graph/graph-node.test.ts
import { beforeEach, describe, expect, test, vi } from 'vitest';
import type { PrintOpts } from '../../../src/grammar-graph/grammar-graph-types.ts';
import { GraphNode, type GraphNodeMeta } from '../../../src/grammar-graph/graph-node.ts';
import { printGraphNode } from '../../../src/grammar-graph/print.ts';
import { RuleRef } from '../../../src/grammar-graph/rule-ref.ts';

vi.mock('../../../src/grammar-graph/print.ts', () => ({
  printGraphNode: vi.fn(() => () => 'mocked_response'),
  printGraphPointer: vi.fn(() => () => 'mocked_response'),
}));

const meta: GraphNodeMeta = { stackId: 1, pathId: 2, stepId: 3 };
const rule = new RuleRef(42);

beforeEach(() => {
  vi.mocked(printGraphNode).mockClear();
});

describe('graph_node', () => {
  test('construct with rule and meta', () => {
    const node = new GraphNode(rule, meta);
    expect(node.rule).toBe(rule);
    expect(node.meta).toBe(meta);
  });

  test('throw if meta is undefined', () => {
    expect(() => new GraphNode(rule, null)).toThrow('Meta is undefined');
  });

  test('correctly calculate and cache its id', () => {
    const node = new GraphNode(rule, meta);
    expect(node.id).toBe('1,2,3');
    node.meta = { stackId: 4, pathId: 5, stepId: 6 };
    expect(node.id).toBe('1,2,3');
  });

  test('delegate print to the print_graph_node function', () => {
    const node = new GraphNode(rule, meta);
    const opts: PrintOpts = { colorize: x => String(x), show_position: false };
    expect(node.print(opts)).toBe('mocked_response');
    expect(printGraphNode).toHaveBeenCalledWith(node);
  });

  test('handle next node linkage', () => {
    const nextNode = new GraphNode(new RuleRef(43), meta);
    const node = new GraphNode(rule, meta, nextNode);
    expect(node.next).toBe(nextNode);
  });
});

File: proposed-projects/round-trip-experiment/reverse/20260909T002206Z_273cf6b9/20260910T042804Z_e72f6c04/ported_implementation/gbnf/grammar_graph/graph_node_test.py
import pytest

from . import graph_node as module
from .grammar_graph_types import PrintOpts
from .graph_node import GraphNode, GraphNodeMeta
from .rule_ref import RuleRef

META = GraphNodeMeta(1, 2, 3)
RULE = RuleRef(42)


def test_construct_with_rule_and_meta():
    node = GraphNode(RULE, META)
    assert node.rule is RULE
    assert node.meta is META


def test_raises_if_meta_is_undefined():
    with pytest.raises(ValueError, match="Meta is undefined"):
        GraphNode(RULE, None)


def test_correctly_calculates_and_caches_its_id():
    node = GraphNode(RULE, META)
    assert node.id == "1,2,3"
    node.meta = GraphNodeMeta(4, 5, 6)
    assert node.id == "1,2,3"


def test_delegates_print_to_the_print_graph_node_function(monkeypatch):
    calls = []

    def mock_print_graph_node(node):
        calls.append(node)
        return lambda opts: "mocked_response"

    monkeypatch.setattr(module, "print_graph_node", mock_print_graph_node)
    node = GraphNode(RULE, META)
    opts = PrintOpts(colorize=lambda text, color: str(text), show_position=False)
    assert node.print(opts) == "mocked_response"
    assert calls == [node]


def test_handles_next_node_linkage():
    next_node = GraphNode(RuleRef(43), META)
    node = GraphNode(RULE, META, next_node)
    assert node.next is next_node
